drgn_tools/cpuinfo.py

```diff
@@ -0,0 +1,615 @@
+                    prog["retbleed_strings"][prog["retbleed_state"]]
+                    .string_()
+                    .decode("utf-8")
+                    + "; SMT "
+                )
+                if not check_smt_enabled(prog):
+                    mitigation += "disabled"
+                elif (
+                    prog["spectre_v2_user_stibp"]
+                    == prog["SPECTRE_V2_USER_STRICT"]
+                    or prog["spectre_v2_user_stibp"]
+                    == prog["SPECTRE_V2_USER_STRICT_PREFERRED"]
+                ):
+                    mitigation += "enabled with STIBP protection"
+                else:
+                    mitigation += "vulnerable"
+        else:
+            mitigation = (
+                prog["retbleed_strings"][prog["retbleed_state"]]
+                .string_()
+                .decode("utf-8")
+            )
+    else:
+        if (
+            prog["retbleed_mitigation"] == prog["RETBLEED_MITIGATION_UNRET"]
+            or prog["retbleed_mitigation"] == prog["RETBLEED_MITIGATION_IBPB"]
+        ):
+            if cpuinfo.x86_vendor != 2 and cpuinfo.x86_vendor != 3:
+                mitigation = "Vulnerable: untrained return thunk / IBPB on non-AMD based uarch"
+            else:
+                mitigation = (
+                    prog["retbleed_strings"][prog["retbleed_mitigation"]]
+                    .string_()
+                    .decode("utf-8")
+                    + "; SMT "
+                )
+                if not check_smt_enabled(prog):
+                    mitigation += "disabled"
+                elif (
+                    prog["spectre_v2_user_stibp"]
+                    == prog["SPECTRE_V2_USER_STRICT"]
+                    or prog["spectre_v2_user_stibp"]
+                    == prog["SPECTRE_V2_USER_STRICT_PREFERRED"]
+                ):
+                    mitigation += "enabled with STIBP protection"
+                else:
+                    mitigation += "vulnerable"
+        else:
+            mitigation = (
+                prog["retbleed_strings"][prog["retbleed_mitigation"]]
+                .string_()
+                .decode("utf-8")
+            )
+
+    return mitigation
+
+
+def get_gds_mitigation(prog: Program) -> str:
+    """
+    Extracts Mitigation for GDS
+    """
+    return (
+        prog["gds_strings"][prog["gds_mitigation"]].string_().decode("utf-8")
+    )
+
+
+def x86_get_cpu_mitigations(prog: Program) -> Dict[str, str]:
+    """
+    Helper to get mitigations for vulnerabilities for x86
+
+    :returns: a dictionary of vulnerabilities with their mitigations
+    """
+    vulns = [
+        "Meltdown",
+        "Spectre_V1",
+        "Spectre_V2",
+        "SSBD",
+        "L1TF",
+        "MDS",
+        "tsx_async_abort",
+        "itlb_multihit",
+        "SRBDS",
+        "mmio_stale_data",
+        "mmio_unknown",
+        "Retbleed",
+        "spec_store_bypass",
+        "GDS",
+        "SRSO",
+    ]
```

Comment:
Let's turn this into a dictionary mapping the vulnerability to the checking function, and add another table mapping the vulnerability to the CPU bug name, if it's different. This way, you don't need that huge if ... elif ... elif block in the loop, you can just lookup the relevant function from the dict.

You'll need to make sure that all of these `get_xxxx_checker()` functions accept two args: `(prog, cpu_caps_bugs)`. A few of them just take `prog`, so you'll need to add a second argument for this to work properly.

Please also create a `get_srso_mitigation()` function that simply returns `"Vulnerable (Status Unknown)"`, so that you don't mix that logic up with this function, which should be simply about running the vulnerability checker for each CPU bug.

The benefit of all of this is that when the next CPU vulnerability happens, we just add a checker, and update each table, and we don't need to make sure we're updating that big if-statement.

```suggestion
    vuln_to_checker = {
        "Meltdown": get_meltdown_mitigation,
        "Spectre_V1": get_spectre_v1_mitigation,
        "Spectre_V2": get_spectre_v2_mitigation,
        "L1TF": get_l1tf_mitigation,
        "MDS": get_mds_mitigation,
        "tsx_async_abort": get_taa_mitigation,
        "itlb_multihit": get_itlb_multihit_mitigation,
        "SRDBS": get_srbds_mitigation,
        "mmio_stale_data": get_mmio_stale_data_mitigation,
        "mmio_unknown": get_mmio_stale_data_mitigation,
        "Retbleed": get_retbleed_mitigation,
        "spec_store_bypass": get_ssb_mitigation,
        "GDS": get_gds_mitigation,
        "SRSO": GET_SRSO_MITIGATION,
        "SSBD": None,
    }
    bug_to_vuln_name = {
        "cpu_meltdown": "Meltdown",
        "spectre_v1": "Spectre_V1",
        "spectre_v2": "Spectre_V2",
        "l1tf": "L1TF",
        "mds": "MDS",
        "taa": "tsx_async_abort",
        "srbds": "SRBDS",
        "retbleed": "Retbleed",
        "gds": "GDS",
        "srso": "SRSO",
    }
```

Reply:
What is the significance of `"SSBD": None` in vuln_to_checker? Because I don't remember encountering any such bug.

Reply:
Your `vulns` list contains `SSBD`, just after `Spectre_V2`. Later on you create the `mitigations` based on the list, so the `mitigations` dictionary you return ends up containing `SSBD: Not Impacted` (since there's no check in the if/elif/elif block that updates it).

If that wasn't intentional, then don't include it there. I was just trying to keep my suggestion identical to yours in logic.

Reply:
Ah I see. It was there in the sample I was seeing in the JIRA, but there were no functions for it in the kernel. I'll remove it. Thanks!